File: python/min_stack.py
import unittest
import math

"""
Design a stack that supports push, pop, top, and retrieving the minimum element 
in constant time.
"""


class MinStack:

    def __init__(self):
        self.s = []
        self.s_min = []

    # @param x, an integer
    def push(self, x):
        self.s.append(x)
        if len(self.s_min) == 0:
            self.s_min.append(x)
        else:
            self.s_min.append(min(self.s_min[-1], x))

    # @return nothing
    def pop(self):
        if len(self.s) == 0:
            return
        self.s.pop()
        self.s_min.pop()

    # @return an integer
    def top(self):
        if len(self.s) == 0:
            return -1

        return self.s[-1]

    # @return an integer
    def getMin(self):
        if len(self.s) == 0:
            return -1

        return self.s_min[-1]


class TestTemplate(unittest.TestCase):

    def test_basic(self):
        s = MinStack()
        s.push(1)
        s.push(2)
        self.assertEqual(1, s.getMin())
        self.assertEqual(2, s.top())
        s.pop()
        self.assertEqual(1, s.getMin())
        self.assertEqual(1, s.top())
        s.pop()

    def test_trivial(self):
        s = MinStack()
        self.assertEqual(-1, s.top())
        s.pop() # Nothing should happen
        self.assertEqual(-1, s.getMin())

    def test_negative(self):
        s = MinStack()
        s.push(-3)
        self.assertEqual(-3, s.getMin())
        s.push(1)
        self.assertEqual(1, s.top())
        self.assertEqual(-3, s.getMin())
        s.push(-2)
        self.assertEqual(-2, s.top())
        self.assertEqual(-3, s.getMin())

    def test_challenging(self):
        s = MinStack()
        s.push(1)
        s.push(2)
        s.push(-2)
        s.push(3)
        s.push(0)
        self.assertEqual(-2, s.getMin())
        s.pop()
        self.assertEqual(-2, s.getMin())
        s.pop()
        self.assertEqual(-2, s.getMin())
        s.pop()
        self.assertEqual(1, s.getMin())
        s.pop()
        self.assertEqual(1, s.getMin())
        s.pop()
        self.assertEqual(-1, s.getMin())


if __name__ == '__main__':
    unittest.main()

File: cpp/MinStack.cpp
#include <gtest/gtest.h>
#include <vector>
#include <math.h>
#include <algorithm>
#include <queue>

class MinStack { 

public:
    MinStack() {
    }

    void push(int x) {
        s.push_back(x);
        if(s_min.empty()){
            s_min.push_back(x);
        } else {
            s_min.push_back(std::min(x, s_min.back()));
        }
    }

    void pop() {
        if(s.empty()){
            return;
        }
        s.pop_back();
        s_min.pop_back();
    }

    int top() {
        if(s.empty()){
            return -1;
        }
        return s.back();
    }

    int getMin() {
        if(s_min.empty()){
            return -1;
        }
        return s_min.back();
    }

private:
    std::vector<int> s;
    std::vector<int> s_min;   
};


TEST(MinStackSuite, StackBasicTests) { 
    std::vector<int> v = {-8, 4, 3, -10, 2, 1};
    std::priority_queue<int, std::vector<int>, std::greater<int> > q;
    MinStack s;

    s.pop(); // should be OK
    ASSERT_EQ(-1 , s.getMin());
    ASSERT_EQ(-1 , s.top());

    for(int i : v){
        s.push(i);
        ASSERT_EQ(s.top(), i);
        q.push(i);
        ASSERT_EQ(s.getMin(), q.top());
    }

    ASSERT_EQ(-10, s.getMin());
    s.pop();
    ASSERT_EQ(-10, s.getMin());
    s.pop();
    ASSERT_EQ(-10, s.getMin());
    s.pop();
    ASSERT_EQ(-8, s.getMin());
    s.pop();
    ASSERT_EQ(-8, s.getMin());
    s.pop();
    s.push(-9);
    ASSERT_EQ(-9, s.getMin());
    s.pop();
    ASSERT_EQ(-8, s.getMin());
}
 
int main(int argc, char **argv) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}